lib/srv/db/proxyserver.go

```diff
@@ -317,11 +327,25 @@ func (s *ProxyServer) Connect(ctx context.Context, user, database string) (net.C
 			}
 			return nil, nil, trace.Wrap(err)
 		}
+
+		// Apply rate limiting.
+		if err := s.cfg.Limiter.AcquireConnection(params.ClientIP); err != nil {
+			return nil, nil, trace.LimitExceeded("client %v exceeded connection limit", params.ClientIP)
+		}
+		// Limiter will be decremented by ConnCloseWrapper below.
+
 		// Upgrade the connection so the client identity can be passed to the
 		// remote server during TLS handshake. On the remote side, the connection
 		// received from the reverse tunnel will be handled by tls.Server.
 		serviceConn = tls.Client(serviceConn, tlsConfig)
-		return serviceConn, proxyContext.authContext, nil
+		// Wrap connection, so we can decrement the limit when the connection is closed.
+		srvConn := &utils.ConnCloseWrapper{
+			Conn: serviceConn,
+			BeforeClose: func() {
+				s.cfg.Limiter.ReleaseConnection(params.ClientIP)
+			},
+		}
```

Comment:
@smallinsky Good catch. I assumed that each connection is closed only one time. I see a few options here:

1. I can move the limiter before the `Connect()` call (`Proxy.HandleConnection()` for each DB), but in this case we would copy the logic for every supported DB
2. I could call `BeforeClose()` only once, but this seems a bit "hacky" to me (also at that point we should also add a mutex)?
3. Move limiter to `Proxy()` function. Looks like a common place too.

Any recommendations?

Reply:
Couple of thoughts:

1. I think that `AcquireConnection` should run _before_ the dial. Otherwise we're not really limiting the connections to remote clusters. Right now it tries to limit after the dial.
2. I agree that it's a bit confusing that we're trying to tie the semaphore release on the downstream connection closure. Maybe `Connect` isn't the best place for it after all. I would probably just do a "acquire" and "defer release" in each ServeXXX method before HandleConnection is invoked. A bit of duplication but would probably be the cleanest way. And we wouldn't need the ConnCloseWrapper either.

@jakule @smallinsky What do you guys think?

Reply:
Agree the AcquireConnection should be called before "dial", though if we place the limit in ServeXXX functions TLS Routing flow will bypass the AcquireConnection call because TLS Router injects db connection directly to the `HandleConnection()`  DB proxy handlers. So first overlapped flow  for separate and multiplex proxy listener mode starts from HandleConnection functions.

Reply:
Ah ok. In that case, I think that ALPN proxy should also accept the same proxy's limiter ([here](https://github.com/gravitational/teleport/blob/v8.0.1/lib/service/service.go#L3157)) and enforce it in its Serve method.

Reply:
There is a risk that proxy limiter won't work properly with ReverseTunnel flow because the handleConnection is invoked asynchronously: https://github.com/gravitational/teleport/blob/master/lib/srv/alpnproxy/proxy.go#L282 

@r0mant @jakule I think that the best place right now to put acquire/release logic is in per proxy DB handleConnection functions like:  https://github.com/gravitational/teleport/blob/master/lib/srv/db/mysql/proxy.go#L55